File: packages/editor-schema/src/markdown.ts
import {
  emptyDoc,
  type DocNode,
  type DocRoot,
  type MarkInstance,
  type MarkName,
} from "./schema.js";
import { validateDoc } from "./validate.js";

/**
 * Bidirectional JSON ↔ Markdown converter for the frozen schema.
 *
 * Design constraints:
 * - Round-trip preservation for the subset of Markdown the schema models —
 *   property-based tests (markdown.test.ts) check that
 *   `markdownToJson(jsonToMarkdown(doc))` equals `doc` (modulo cosmetic
 *   whitespace) for any well-formed doc.
 * - No external Markdown parser — the surface we need is small (headings,
 *   lists, code, images, attachments, internal links, hard breaks) and a
 *   bespoke parser is easier to keep in sync with the Kotlin twin than a
 *   third-party parser.
 */

// ---------- JSON → Markdown ----------

export function jsonToMarkdown(doc: DocRoot): string {
  validateDoc(doc);
  const lines: string[] = [];
  const content = doc.content ?? [];
  for (let i = 0; i < content.length; i += 1) {
    const block = content[i] as DocNode;
    lines.push(blockToMarkdown(block));
    if (i < content.length - 1) lines.push("");
  }
  return lines.join("\n");
}

function blockToMarkdown(node: DocNode): string {
  switch (node.type) {
    case "paragraph":
      return inlineToMarkdown(node.content ?? []);
    case "heading": {
      const level = (node.attrs as { level: 1 | 2 | 3 }).level;
      return `${"#".repeat(level)} ${inlineToMarkdown(node.content ?? [])}`;
    }
    case "bullet_list":
      return listToMarkdown(node, () => "- ");
    case "ordered_list":
      return listToMarkdown(node, (i) => `${i + 1}. `);
    case "check_list":
      return checkListToMarkdown(node);
    case "code_block":
      return codeBlockToMarkdown(node);
    case "image": {
      const attrs = node.attrs as { src: string; alt?: string };
      return `![${attrs.alt ?? ""}](${attrs.src})`;
    }
    case "attachment": {
      const attrs = node.attrs as { attachmentId: string; displayName: string };
      return `[[attachment:${attrs.attachmentId}|${escapeAttachmentLabel(attrs.displayName)}]]`;
    }
    default:
      throw new Error(`Cannot serialize block ${JSON.stringify(node.type)}`);
  }
}

function listToMarkdown(node: DocNode, prefix: (index: number) => string): string {
  const items = node.content ?? [];
  return items
    .map((item, idx) => {
      const itemContent = (item.content ?? []).map((b) => blockToMarkdown(b)).join("\n");
      return indentContinuation(`${prefix(idx)}${itemContent}`, prefix(idx).length);
    })
    .join("\n");
}

function checkListToMarkdown(node: DocNode): string {
  const items = node.content ?? [];
  return items
    .map((item) => {
      const checked = (item.attrs as { checked: boolean }).checked;
      const itemContent = (item.content ?? []).map((b) => blockToMarkdown(b)).join("\n");
      const marker = checked ? "- [x] " : "- [ ] ";
      return indentContinuation(`${marker}${itemContent}`, marker.length);
    })
    .join("\n");
}

function indentContinuation(text: string, indent: number): string {
  const pad = " ".repeat(indent);
  const [first = "", ...rest] = text.split("\n");
  if (rest.length === 0) return first;
  return [first, ...rest.map((l) => `${pad}${l}`)].join("\n");
}

function codeBlockToMarkdown(node: DocNode): string {
  const lang = (node.attrs as { language?: string | null } | undefined)?.language ?? "";
  const body = (node.content ?? []).map((c) => (typeof c.text === "string" ? c.text : "")).join("");
  return `\`\`\`${lang}\n${body}\n\`\`\``;
}

function inlineToMarkdown(inline: readonly DocNode[]): string {
  return inline.map(inlineNodeToMarkdown).join("");
}

function inlineNodeToMarkdown(node: DocNode): string {
  switch (node.type) {
    case "text":
      return applyMarks(escapeInline(node.text ?? ""), node.marks ?? []);
    case "code_inline":
      // Backticks inside inline code: pick the shortest run that won't clash.
      return wrapInlineCode(node.text ?? "");
    case "internal_link": {
      const attrs = node.attrs as { noteId: string; label?: string };
      const label = attrs.label !== undefined && attrs.label.length > 0 ? `|${attrs.label}` : "";
      return `[[note:${attrs.noteId}${label}]]`;
    }
    case "hard_break":
      return "  \n";
    default:
      throw new Error(`Cannot serialize inline ${JSON.stringify(node.type)}`);
  }
}

const MARK_ORDER: readonly MarkName[] = ["bold", "italic", "strike"];
const MARK_DELIM: Record<MarkName, string> = {
  bold: "**",
  italic: "_",
  strike: "~~",
};

function applyMarks(text: string, marks: readonly MarkInstance[]): string {
  const ordered = [...marks].sort(
    (a, b) => MARK_ORDER.indexOf(a.type) - MARK_ORDER.indexOf(b.type),
  );
  let out = text;
  for (const m of ordered) {
    const d = MARK_DELIM[m.type];
    out = `${d}${out}${d}`;
  }
  return out;
}

function wrapInlineCode(text: string): string {
  let runs = 1;
  const matches = text.match(/`+/g);
  if (matches !== null) {
    runs = Math.max(...matches.map((m) => m.length)) + 1;
  }
  const fence = "`".repeat(runs);
  // Pad with a space when the content starts/ends with a backtick to keep
  // the fence unambiguous.
  const padded = text.startsWith("`") || text.endsWith("`") ? ` ${text} ` : text;
  return `${fence}${padded}${fence}`;
}

function escapeInline(text: string): string {
  // Escape Markdown-significant characters in plain text.
  return text.replace(/([\\`*_~\[\]])/g, "\\$1");
}

function escapeAttachmentLabel(label: string): string {
  return label.replace(/\|/g, "\\|").replace(/]]/g, "\\]\\]");
}

// ---------- Markdown → JSON ----------

const HEADING_RE = /^(#{1,3})\s+(.*)$/;
const FENCE_RE = /^```(\S*)\s*$/;
const BULLET_RE = /^(\s*)[-*]\s+(.*)$/;
const ORDERED_RE = /^(\s*)(\d+)\.\s+(.*)$/;
const CHECK_RE = /^(\s*)[-*]\s+\[( |x|X)\]\s+(.*)$/;
const ATTACHMENT_BLOCK_RE = /^\[\[attachment:([^|\]]+)\|((?:\\.|[^\]])+)\]\]$/;
const IMAGE_BLOCK_RE = /^!\[([^\]]*)\]\(([^)]+)\)$/;

interface Cursor {
  readonly lines: readonly string[];
  index: number;
}

export function markdownToJson(markdown: string): DocRoot {
  const cursor: Cursor = { lines: markdown.split("\n"), index: 0 };
  const blocks: DocNode[] = [];
  while (cursor.index < cursor.lines.length) {
    const line = (cursor.lines[cursor.index] ?? "").replace(/\s+$/, "");
    if (line.length === 0) {
      cursor.index += 1;
      continue;
    }
    blocks.push(parseBlock(cursor));
  }
  const doc: DocRoot = blocks.length === 0 ? emptyDoc() : { type: "doc", content: blocks };
  validateDoc(doc);
  return doc;
}

function parseBlock(cursor: Cursor): DocNode {
  const line = (cursor.lines[cursor.index] ?? "").replace(/\s+$/, "");
  const headingMatch = HEADING_RE.exec(line);
  if (headingMatch !== null) {
    cursor.index += 1;
    return {
      type: "heading",
      attrs: { level: headingMatch[1]!.length as 1 | 2 | 3 },
      content: parseInline(headingMatch[2] ?? ""),
    };
  }
  if (FENCE_RE.test(line)) {
    return parseCodeBlock(cursor);
  }
  if (IMAGE_BLOCK_RE.test(line)) {
    cursor.index += 1;
    const m = IMAGE_BLOCK_RE.exec(line)!;
    const node: DocNode = {
      type: "image",
      attrs: { src: m[2]!, ...(m[1] !== "" ? { alt: m[1]! } : {}) },
    };
    return node;
  }
  if (ATTACHMENT_BLOCK_RE.test(line)) {
    cursor.index += 1;
    const m = ATTACHMENT_BLOCK_RE.exec(line)!;
    return {
      type: "attachment",
      attrs: {
        attachmentId: m[1]!,
        displayName: unescapeAttachmentLabel(m[2]!),
      },
    };
  }
  if (CHECK_RE.test(line)) {
    return parseList(cursor, "check_list");
  }
  if (BULLET_RE.test(line)) {
    return parseList(cursor, "bullet_list");
  }
  if (ORDERED_RE.test(line)) {
    return parseList(cursor, "ordered_list");
  }
  return parseParagraph(cursor);
}

function parseCodeBlock(cursor: Cursor): DocNode {
  const fenceLine = (cursor.lines[cursor.index] ?? "").trim();
  const lang = FENCE_RE.exec(fenceLine)?.[1] ?? "";
  cursor.index += 1;
  const bodyLines: string[] = [];
  while (cursor.index < cursor.lines.length) {
    const l = cursor.lines[cursor.index] ?? "";
    if (l.trimEnd() === "```") {
      cursor.index += 1;
      break;
    }
    bodyLines.push(l);
    cursor.index += 1;
  }
  const text = bodyLines.join("\n");
  const node: DocNode = {
    type: "code_block",
    attrs: { language: lang.length === 0 ? null : lang },
    content: text.length === 0 ? [] : [{ type: "text", text }],
  };
  return node;
}

function parseList(cursor: Cursor, kind: "bullet_list" | "ordered_list" | "check_list"): DocNode {
  const items: DocNode[] = [];
  while (cursor.index < cursor.lines.length) {
    const line = (cursor.lines[cursor.index] ?? "").replace(/\s+$/, "");
    if (line.length === 0) break;
    const detected = detectListKind(line);
    if (detected !== kind) break;
    items.push(parseListItem(cursor, kind));
  }
  return { type: kind, content: items };
}

function detectListKind(line: string): "bullet_list" | "ordered_list" | "check_list" | null {
  if (CHECK_RE.test(line)) return "check_list";
  if (BULLET_RE.test(line)) return "bullet_list";
  if (ORDERED_RE.test(line)) return "ordered_list";
  return null;
}

function parseListItem(
  cursor: Cursor,
  kind: "bullet_list" | "ordered_list" | "check_list",
): DocNode {
  const line = (cursor.lines[cursor.index] ?? "").replace(/\s+$/, "");
  let rest: string;
  let checked: boolean | null = null;
  if (kind === "check_list") {
    const m = CHECK_RE.exec(line)!;
    checked = m[2]?.toLowerCase() === "x";
    rest = m[3] ?? "";
  } else if (kind === "bullet_list") {
    const m = BULLET_RE.exec(line)!;
    rest = m[2] ?? "";
  } else {
    const m = ORDERED_RE.exec(line)!;
    rest = m[3] ?? "";
  }
  cursor.index += 1;
  // Collect continuation lines (indented).
  const paragraphLines: string[] = [rest];
  while (cursor.index < cursor.lines.length) {
    const next = cursor.lines[cursor.index] ?? "";
    if (next.length === 0) break;
    if (detectListKind(next.trimStart()) !== null && next.startsWith(" ") === false) break;
    if (next.startsWith(" ")) {
      paragraphLines.push(next.replace(/^\s+/, ""));
      cursor.index += 1;
    } else {
      break;
    }
  }
  const item: DocNode = {
    type: kind === "check_list" ? "check_list_item" : "list_item",
    ...(kind === "check_list" ? { attrs: { checked: checked ?? false } } : {}),
    content: [{ type: "paragraph", content: parseInline(paragraphLines.join("\n")) }],
  };
  return item;
}

function parseParagraph(cursor: Cursor): DocNode {
  const collected: string[] = [];
  while (cursor.index < cursor.lines.length) {
    const line = (cursor.lines[cursor.index] ?? "").replace(/\s+$/, "");
    if (line.length === 0) break;
    if (
      HEADING_RE.test(line) ||
      FENCE_RE.test(line) ||
      detectListKind(line) !== null ||
      IMAGE_BLOCK_RE.test(line) ||
      ATTACHMENT_BLOCK_RE.test(line)
    ) {
      break;
    }
    collected.push(cursor.lines[cursor.index] ?? "");
    cursor.index += 1;
  }
  return { type: "paragraph", content: parseInline(collected.join("\n")) };
}

// Inline parsing — small, deterministic, intentionally not Markdown-spec
// strict. We only emit / consume what we ourselves produce; user input is
// always tunnelled through the document model.
function parseInline(text: string): DocNode[] {
  if (text.length === 0) return [];
  const nodes: DocNode[] = [];
  let pos = 0;
  let buffer = "";
  const flushText = (marks: MarkName[] = []) => {
    if (buffer.length === 0) return;
    nodes.push({
      type: "text",
      text: buffer,
      ...(marks.length > 0 ? { marks: marks.map((type) => ({ type })) } : {}),
    });
    buffer = "";
  };

  while (pos < text.length) {
    const c = text[pos]!;
    // Hard break: two trailing spaces before a newline.
    if (c === " " && text[pos + 1] === " " && text[pos + 2] === "\n") {
      flushText();
      nodes.push({ type: "hard_break" });
      pos += 3;
      continue;
    }
    if (c === "\\" && pos + 1 < text.length) {
      buffer += text[pos + 1]!;
      pos += 2;
      continue;
    }
    if (c === "`") {
      flushText();
      const closing = text.indexOf("`", pos + 1);
      if (closing === -1) {
        buffer += c;
        pos += 1;
        continue;
      }
      const raw = text.slice(pos + 1, closing);
      // Strip the optional padding space iff the serializer would have
      // added it — i.e. the inner content has a backtick adjacent to the
      // padding. Pure whitespace content is preserved verbatim.
      const wasPadded =
        raw.length >= 2 &&
        raw.startsWith(" ") &&
        raw.endsWith(" ") &&
        (raw[1] === "`" || raw[raw.length - 2] === "`");
      const trimmed = wasPadded ? raw.slice(1, -1) : raw;
      nodes.push({ type: "code_inline", text: trimmed });
      pos = closing + 1;
      continue;
    }
    if (c === "[" && text[pos + 1] === "[") {
      const end = text.indexOf("]]", pos + 2);
      if (end === -1) {
        buffer += c;
        pos += 1;
        continue;
      }
      const body = text.slice(pos + 2, end);
      const noteMatch = /^note:([^|]+)(?:\|(.*))?$/.exec(body);
      if (noteMatch !== null) {
        flushText();
        nodes.push({
          type: "internal_link",
          attrs: {
            noteId: noteMatch[1]!,
            ...(noteMatch[2] !== undefined && noteMatch[2].length > 0
              ? { label: noteMatch[2] }
              : {}),
          },
        });
        pos = end + 2;
        continue;
      }
      buffer += c;
      pos += 1;
      continue;
    }
    if (c === "*" && text[pos + 1] === "*") {
      const close = findMarkClose(text, pos + 2, "**");
      if (close !== -1) {
        flushText();
        nodes.push(...parseInlineMarked(text.slice(pos + 2, close), "bold"));
        pos = close + 2;
        continue;
      }
    }
    if (c === "_") {
      const close = findMarkClose(text, pos + 1, "_");
      if (close !== -1) {
        flushText();
        nodes.push(...parseInlineMarked(text.slice(pos + 1, close), "italic"));
        pos = close + 1;
        continue;
      }
    }
    if (c === "~" && text[pos + 1] === "~") {
      const close = findMarkClose(text, pos + 2, "~~");
      if (close !== -1) {
        flushText();
        nodes.push(...parseInlineMarked(text.slice(pos + 2, close), "strike"));
        pos = close + 2;
        continue;
      }
    }
    buffer += c;
    pos += 1;
  }
  flushText();
  return nodes;
}

function findMarkClose(text: string, start: number, delim: string): number {
  let i = start;
  while (i < text.length) {
    if (text[i] === "\\") {
      i += 2;
      continue;
    }
    if (text.slice(i, i + delim.length) === delim) return i;
    i += 1;
  }
  return -1;
}

function parseInlineMarked(text: string, mark: MarkName): DocNode[] {
  const inner = parseInline(text);
  return inner.map((node) => {
    if (node.type !== "text" && node.type !== "code_inline") return node;
    const existing = node.marks ?? [];
    if (existing.some((m) => m.type === mark)) return node;
    return { ...node, marks: [...existing, { type: mark }] };
  });
}

function unescapeAttachmentLabel(label: string): string {
  return label.replace(/\\\]\\\]/g, "]]").replace(/\\\|/g, "|");
}

File: packages/android/editor-schema/src/main/kotlin/app/notes/editor/Markdown.kt
package app.notes.editor

/**
 * Mirror of `markdown.ts`. Serializes/parses the subset of Markdown the
 * frozen schema produces. The two implementations are kept in lockstep by
 * the round-trip property tests on both sides.
 */

private val MARK_ORDER = listOf(MarkName.BOLD, MarkName.ITALIC, MarkName.STRIKE)
private val MARK_DELIM = mapOf(
    MarkName.BOLD to "**",
    MarkName.ITALIC to "_",
    MarkName.STRIKE to "~~",
)

// ---------- JSON → Markdown ----------

fun jsonToMarkdown(doc: Doc): String {
    validateDoc(doc)
    return doc.content.joinToString("\n\n") { blockToMarkdown(it) }
}

private fun blockToMarkdown(node: DocNode): String = when (BlockNodeName.fromWire(node.type)!!) {
    BlockNodeName.PARAGRAPH -> inlineToMarkdown(node.content)
    BlockNodeName.HEADING -> {
        val level = node.attrs["level"] as Int
        "${"#".repeat(level)} ${inlineToMarkdown(node.content)}"
    }
    BlockNodeName.BULLET_LIST -> listToMarkdown(node) { "- " }
    BlockNodeName.ORDERED_LIST -> listToMarkdown(node) { idx -> "${idx + 1}. " }
    BlockNodeName.CHECK_LIST -> checkListToMarkdown(node)
    BlockNodeName.CODE_BLOCK -> codeBlockToMarkdown(node)
    BlockNodeName.IMAGE -> {
        val src = node.attrs["src"] as String
        val alt = node.attrs["alt"] as? String ?: ""
        "![${alt}](${src})"
    }
    BlockNodeName.ATTACHMENT -> {
        val id = node.attrs["attachmentId"] as String
        val name = node.attrs["displayName"] as String
        "[[attachment:$id|${escapeAttachmentLabel(name)}]]"
    }
}

private fun listToMarkdown(node: DocNode, prefix: (Int) -> String): String =
    node.content.mapIndexed { idx, item ->
        val itemBody = item.content.joinToString("\n") { blockToMarkdown(it) }
        indentContinuation("${prefix(idx)}$itemBody", prefix(idx).length)
    }.joinToString("\n")

private fun checkListToMarkdown(node: DocNode): String =
    node.content.joinToString("\n") { item ->
        val checked = item.attrs["checked"] as Boolean
        val marker = if (checked) "- [x] " else "- [ ] "
        val itemBody = item.content.joinToString("\n") { blockToMarkdown(it) }
        indentContinuation("$marker$itemBody", marker.length)
    }

private fun indentContinuation(text: String, indent: Int): String {
    val parts = text.split("\n")
    if (parts.size == 1) return parts[0]
    val pad = " ".repeat(indent)
    return parts.first() + "\n" + parts.drop(1).joinToString("\n") { "$pad$it" }
}

private fun codeBlockToMarkdown(node: DocNode): String {
    val lang = node.attrs["language"] as? String ?: ""
    val body = node.content.joinToString("") { it.text ?: "" }
    return "```$lang\n$body\n```"
}

private fun inlineToMarkdown(inline: List<DocNode>): String =
    inline.joinToString("") { inlineNodeToMarkdown(it) }

private fun inlineNodeToMarkdown(node: DocNode): String =
    when (InlineNodeName.fromWire(node.type)!!) {
        InlineNodeName.TEXT -> applyMarks(escapeInline(node.text ?: ""), node.marks)
        InlineNodeName.CODE_INLINE -> wrapInlineCode(node.text ?: "")
        InlineNodeName.INTERNAL_LINK -> {
            val id = node.attrs["noteId"] as String
            val label = node.attrs["label"] as? String
            if (!label.isNullOrEmpty()) "[[note:$id|$label]]" else "[[note:$id]]"
        }
        InlineNodeName.HARD_BREAK -> "  \n"
    }

private fun applyMarks(text: String, marks: List<Mark>): String {
    if (marks.isEmpty()) return text
    val ordered = marks.sortedBy { MARK_ORDER.indexOf(it.type) }
    var out = text
    for (m in ordered) {
        val d = MARK_DELIM.getValue(m.type)
        out = "$d$out$d"
    }
    return out
}

private fun wrapInlineCode(text: String): String {
    val runs = Regex("`+").findAll(text).maxOfOrNull { it.value.length } ?: 0
    val fence = "`".repeat(runs + 1)
    val padded = if (text.startsWith("`") || text.endsWith("`")) " $text " else text
    return "$fence$padded$fence"
}

private fun escapeInline(text: String): String =
    text.replace(Regex("([\\\\`*_~\\[\\]])"), "\\\\$1")

private fun escapeAttachmentLabel(label: String): String =
    label.replace("|", "\\|").replace("]]", "\\]\\]")

// ---------- Markdown → JSON ----------

private val HEADING_RE = Regex("^(#{1,3})\\s+(.*)$")
private val FENCE_RE = Regex("^```(\\S*)\\s*$")
private val BULLET_RE = Regex("^(\\s*)[-*]\\s+(.*)$")
private val ORDERED_RE = Regex("^(\\s*)(\\d+)\\.\\s+(.*)$")
private val CHECK_RE = Regex("^(\\s*)[-*]\\s+\\[( |x|X)]\\s+(.*)$")
private val ATTACHMENT_BLOCK_RE = Regex("^\\[\\[attachment:([^|\\]]+)\\|((?:\\\\.|[^\\]])+)]]$")
private val IMAGE_BLOCK_RE = Regex("^!\\[([^\\]]*)]\\(([^)]+)\\)$")

private class Cursor(val lines: List<String>) {
    var index: Int = 0
}

fun markdownToJson(markdown: String): Doc {
    val cursor = Cursor(markdown.split("\n"))
    val blocks = mutableListOf<DocNode>()
    while (cursor.index < cursor.lines.size) {
        val line = (cursor.lines[cursor.index]).trimEnd()
        if (line.isEmpty()) {
            cursor.index += 1
            continue
        }
        blocks.add(parseBlock(cursor))
    }
    val doc = if (blocks.isEmpty()) Doc.empty() else Doc(blocks)
    validateDoc(doc)
    return doc
}

private fun parseBlock(cursor: Cursor): DocNode {
    val line = cursor.lines[cursor.index].trimEnd()
    HEADING_RE.matchEntire(line)?.let { m ->
        cursor.index += 1
        return DocNode(
            type = "heading",
            attrs = mapOf("level" to m.groupValues[1].length),
            content = parseInline(m.groupValues[2]),
        )
    }
    if (FENCE_RE.matches(line)) return parseCodeBlock(cursor)
    IMAGE_BLOCK_RE.matchEntire(line)?.let { m ->
        cursor.index += 1
        val attrs = mutableMapOf<String, Any?>("src" to m.groupValues[2])
        if (m.groupValues[1].isNotEmpty()) attrs["alt"] = m.groupValues[1]
        return DocNode(type = "image", attrs = attrs)
    }
    ATTACHMENT_BLOCK_RE.matchEntire(line)?.let { m ->
        cursor.index += 1
        return DocNode(
            type = "attachment",
            attrs = mapOf(
                "attachmentId" to m.groupValues[1],
                "displayName" to unescapeAttachmentLabel(m.groupValues[2]),
            ),
        )
    }
    if (CHECK_RE.matches(line)) return parseList(cursor, "check_list")
    if (BULLET_RE.matches(line)) return parseList(cursor, "bullet_list")
    if (ORDERED_RE.matches(line)) return parseList(cursor, "ordered_list")
    return parseParagraph(cursor)
}

private fun parseCodeBlock(cursor: Cursor): DocNode {
    val fence = cursor.lines[cursor.index].trim()
    val lang = FENCE_RE.matchEntire(fence)?.groupValues?.get(1).orEmpty()
    cursor.index += 1
    val body = mutableListOf<String>()
    while (cursor.index < cursor.lines.size) {
        val l = cursor.lines[cursor.index]
        if (l.trimEnd() == "```") {
            cursor.index += 1
            break
        }
        body.add(l)
        cursor.index += 1
    }
    val text = body.joinToString("\n")
    return DocNode(
        type = "code_block",
        attrs = mapOf("language" to if (lang.isEmpty()) null else lang),
        content = if (text.isEmpty()) emptyList() else listOf(DocNode(type = "text", text = text)),
    )
}

private fun detectListKind(line: String): String? = when {
    CHECK_RE.matches(line) -> "check_list"
    BULLET_RE.matches(line) -> "bullet_list"
    ORDERED_RE.matches(line) -> "ordered_list"
    else -> null
}

private fun parseList(cursor: Cursor, kind: String): DocNode {
    val items = mutableListOf<DocNode>()
    while (cursor.index < cursor.lines.size) {
        val line = cursor.lines[cursor.index].trimEnd()
        if (line.isEmpty()) break
        if (detectListKind(line) != kind) break
        items.add(parseListItem(cursor, kind))
    }
    return DocNode(type = kind, content = items)
}

private fun parseListItem(cursor: Cursor, kind: String): DocNode {
    val line = cursor.lines[cursor.index].trimEnd()
    val (rest, checked) = when (kind) {
        "check_list" -> {
            val m = CHECK_RE.matchEntire(line)!!
            m.groupValues[3] to (m.groupValues[2].equals("x", ignoreCase = true))
        }
        "bullet_list" -> BULLET_RE.matchEntire(line)!!.groupValues[2] to null
        else -> ORDERED_RE.matchEntire(line)!!.groupValues[3] to null
    }
    cursor.index += 1
    val collected = mutableListOf(rest)
    while (cursor.index < cursor.lines.size) {
        val next = cursor.lines[cursor.index]
        if (next.isEmpty()) break
        if (detectListKind(next.trimStart()) != null && !next.startsWith(" ")) break
        if (next.startsWith(" ")) {
            collected.add(next.trimStart())
            cursor.index += 1
        } else break
    }
    val paragraph = DocNode(type = "paragraph", content = parseInline(collected.joinToString("\n")))
    val itemType = if (kind == "check_list") "check_list_item" else "list_item"
    val attrs: Map<String, Any?> = if (checked != null) mapOf("checked" to checked) else emptyMap()
    return DocNode(type = itemType, attrs = attrs, content = listOf(paragraph))
}

private fun parseParagraph(cursor: Cursor): DocNode {
    val collected = mutableListOf<String>()
    while (cursor.index < cursor.lines.size) {
        val line = cursor.lines[cursor.index].trimEnd()
        if (line.isEmpty()) break
        if (
            HEADING_RE.matches(line) ||
            FENCE_RE.matches(line) ||
            detectListKind(line) != null ||
            IMAGE_BLOCK_RE.matches(line) ||
            ATTACHMENT_BLOCK_RE.matches(line)
        ) break
        collected.add(cursor.lines[cursor.index])
        cursor.index += 1
    }
    return DocNode(type = "paragraph", content = parseInline(collected.joinToString("\n")))
}

private fun parseInline(text: String): List<DocNode> {
    if (text.isEmpty()) return emptyList()
    val nodes = mutableListOf<DocNode>()
    val buffer = StringBuilder()
    fun flush(marks: List<MarkName> = emptyList()) {
        if (buffer.isEmpty()) return
        nodes.add(
            DocNode(
                type = "text",
                text = buffer.toString(),
                marks = marks.map { Mark(it) },
            ),
        )
        buffer.clear()
    }
    var pos = 0
    while (pos < text.length) {
        val c = text[pos]
        // Hard break: two spaces before \n.
        if (c == ' ' && pos + 2 < text.length && text[pos + 1] == ' ' && text[pos + 2] == '\n') {
            flush()
            nodes.add(DocNode(type = "hard_break"))
            pos += 3
            continue
        }
        if (c == '\\' && pos + 1 < text.length) {
            buffer.append(text[pos + 1])
            pos += 2
            continue
        }
        if (c == '`') {
            flush()
            val closing = text.indexOf('`', pos + 1)
            if (closing == -1) {
                buffer.append(c)
                pos += 1
                continue
            }
            val raw = text.substring(pos + 1, closing)
            val padded = raw.length >= 2 && raw.startsWith(" ") && raw.endsWith(" ") &&
                (raw[1] == '`' || raw[raw.length - 2] == '`')
            val trimmed = if (padded) raw.substring(1, raw.length - 1) else raw
            nodes.add(DocNode(type = "code_inline", text = trimmed))
            pos = closing + 1
            continue
        }
        if (c == '[' && pos + 1 < text.length && text[pos + 1] == '[') {
            val end = text.indexOf("]]", pos + 2)
            if (end == -1) {
                buffer.append(c)
                pos += 1
                continue
            }
            val body = text.substring(pos + 2, end)
            val noteMatch = Regex("^note:([^|]+)(?:\\|(.*))?$").matchEntire(body)
            if (noteMatch != null) {
                flush()
                val attrs = mutableMapOf<String, Any?>("noteId" to noteMatch.groupValues[1])
                val label = noteMatch.groupValues[2]
                if (label.isNotEmpty()) attrs["label"] = label
                nodes.add(DocNode(type = "internal_link", attrs = attrs))
                pos = end + 2
                continue
            }
            buffer.append(c)
            pos += 1
            continue
        }
        if (c == '*' && pos + 1 < text.length && text[pos + 1] == '*') {
            val close = findMarkClose(text, pos + 2, "**")
            if (close != -1) {
                flush()
                nodes.addAll(parseInlineMarked(text.substring(pos + 2, close), MarkName.BOLD))
                pos = close + 2
                continue
            }
        }
        if (c == '_') {
            val close = findMarkClose(text, pos + 1, "_")
            if (close != -1) {
                flush()
                nodes.addAll(parseInlineMarked(text.substring(pos + 1, close), MarkName.ITALIC))
                pos = close + 1
                continue
            }
        }
        if (c == '~' && pos + 1 < text.length && text[pos + 1] == '~') {
            val close = findMarkClose(text, pos + 2, "~~")
            if (close != -1) {
                flush()
                nodes.addAll(parseInlineMarked(text.substring(pos + 2, close), MarkName.STRIKE))
                pos = close + 2
                continue
            }
        }
        buffer.append(c)
        pos += 1
    }
    flush()
    return nodes
}

private fun findMarkClose(text: String, start: Int, delim: String): Int {
    var i = start
    while (i < text.length) {
        if (text[i] == '\\') {
            i += 2
            continue
        }
        if (i + delim.length <= text.length && text.substring(i, i + delim.length) == delim) return i
        i += 1
    }
    return -1
}

private fun parseInlineMarked(text: String, mark: MarkName): List<DocNode> {
    val inner = parseInline(text)
    return inner.map { node ->
        if (node.type != "text" && node.type != "code_inline") return@map node
        if (node.marks.any { it.type == mark }) return@map node
        node.copy(marks = node.marks + Mark(mark))
    }
}

private fun unescapeAttachmentLabel(label: String): String =
    label.replace("\\]\\]", "]]").replace("\\|", "|")
